rust/src/contract_utils/entrypoint.rs

```diff
@@ -0,0 +1,133 @@
+- e.g. if the execution network will store the state of the contracts - as the WASM contract module memory
+- it would allow to read other contract's state "directly" from WASM module memory.
+*/
+
+// inspired by https://github.com/dfinity/examples/blob/master/rust/basic_dao/src/basic_dao/src/lib.rs#L13
+thread_local! {
+    static STATE: RefCell<State> = RefCell::default();
+}
+
+#[wasm_bindgen()]
+pub async fn handle_entrypoint(interaction: JsValue) -> Option<JsValue> {
+    log_tx();
+
+    let result: Result<HandlerResult, ContractError>;
+    let action: Result<Action, Error> = interaction.into_serde();
+
+    if action.is_err() {
+        // cannot pass any data from action.error here - ends up with
+        // "FnOnce called more than once" error from wasm-bindgen for
+        // "foreign_call" testcase.
+        result = Err(ContractError::RuntimeError("Error while parsing input".to_string()));
+    } else {
+        // not sure about clone here
+        let current_state = STATE.with(|service| service.borrow().clone());
+
+        result = handle(current_state, action.unwrap()).await;
+    }
+
+    if result.is_ok() {
+        log("Result ok");
+        let handler_result = result.as_ref().ok().unwrap();
+
+        return if let HandlerResult::NewState(state) = handler_result {
+            // not sure about clone here
+            STATE.with(|service| service.replace(state.clone()));
+            None
+        } else {
+            Some(JsValue::from_serde(&result).unwrap())
+        };
+    } else {
+        Some(JsValue::from_serde(&result).unwrap())
+    }
+}
+
+fn log_tx() {
+    log(&format!("Block indep_hash {}", Block::indep_hash()));
+    log(&format!("Block height {}", Block::height()));
+    log(&format!("Block timestamp {}", Block::timestamp()));
+
+    log(&format!("Contract id {}", Contract::id()));
+    log(&format!("Contract owner {}", Contract::owner()));
+
+    log(&format!("Transaction id {}", Transaction::id()));
+    log(&format!("Transaction owner {}", Transaction::owner()));
+    log(&format!("Transaction target {}", Transaction::target()));
+}
+
+
+#[wasm_bindgen(js_name = initState)]
+pub fn init_state(state: &JsValue) {
+    log(&format!("Calling init state 2"));
+    let state_parsed: State = state.into_serde().unwrap();
+    log(&format!("state parsed"));
+
+    STATE.with(|service| service.replace(state_parsed));
+}
+
+#[wasm_bindgen(js_name = currentState)]
+pub fn current_state() -> JsValue {
+    log(&format!("Calling current state"));
+
+    // not sure if that's deterministic - which is very important for the execution network.
+    // TODO: perf - according to docs:
+    // "This is unlikely to be super speedy so it's not recommended for large payload"
+    // - we should minimize calls to from_serde
+    let current_state = STATE.with(|service| service.borrow().clone());
+    JsValue::from_serde(&current_state).unwrap()
+}
+
+#[wasm_bindgen()]
+pub fn lang() -> String {
```

Comment:
this function and the next one (`contract_type`) needs to be adopted to the latest changes in the SDK.
https://github.com/redstone-finance/redstone-contracts-wasm/commit/7920c50da1d56a6c1e617a0fc5372b53dff58891#diff-f79aada6c062906719b8c4ea9c22fb916d82fc34d72fb0ccd80190bd80355ce2